src/appleseed/foundation/array/arrayref.h

```diff
@@ -0,0 +1,182 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#pragma once
+
+// appleseed.foundation headers.
+#include "foundation/array/array.h"
+#include "foundation/array/arrayview.h"
+#include "foundation/platform/compiler.h"
+
+// Standard headers.
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#if APPLESEED_COMPILER_CXX_GENERALIZED_INITIALIZERS
+#include <initializer_list>
+#endif
+
+namespace foundation
+{
+
+//
+// Typed non const reference to an array.
+//
+
+template <typename T>
+class ArrayRef
+{
+  public:
+    typedef T value_type;
+    typedef value_type& reference;
+    typedef const value_type& const_reference;
+    typedef size_t size_type;
+
+    // Constructor.
+    explicit ArrayRef(Array& array)
+      : m_array(const_cast<Array&>(array))
+    {
+    }
+
+    // Implicit conversion to ArrayView.
+    operator const ArrayView<T>() const
+    {
+        return ArrayView<T>(m_array);
+    }
+
+    // Return true if the array is empty.
+    bool empty() const
+    {
+        return m_array.empty();
+    }
+
+    // Return the number of items in the array.
+    size_t size() const
+    {
+        return m_array.size();
+    }
+
+    // Return a pointer to the beginning of the array.
+    const T* begin() const
+    {
+        return reinterpret_cast<const T*>(m_array.begin());
+    }
+
+    // Return a pointer to the end of the array.
+    const T* end() const
+    {
+        return reinterpret_cast<const T*>(m_array.end());
+    }
+
+    // Return a pointer to the beginning of the array.
+    T* begin()
+    {
+        return reinterpret_cast<T*>(m_array.begin());
+    }
+
+    // Return a pointer to the end of the array.
+    T* end()
+    {
+        return reinterpret_cast<T*>(m_array.end());
+    }
+
+    // Array element access.
+    const T& operator[](const size_t i) const
+    {
+        assert(i < size());
+        return begin()[i];
+    }
+
+    T& operator[](const size_t i)
+    {
+        assert(i < this->size());
+        return begin()[i];
+    }
+
+    // Remove all items from the array.
+    void clear()
+    {
+        this->m_array.clear();
+    }
+
+    // Return the capacity of the array.
+    size_t capacity() const
+    {
+        return m_array.capacity();
+    }
+
+    // Reserve memory for n items.
+    void reserve(const size_t n)
+    {
+        m_array.reserve(n);
+    }
+
+    // Resize the array to n items.
+    void resize(const size_t n)
+    {
+        m_array.resize(n);
+    }
+
+    // Remove excess capacity from the array.
+    void shrink_to_fit()
+    {
+        m_array.shrink_to_fit();
+    }
+
+    // Add new items to the array.
+    void push_back(const T& x)
+    {
+        m_array.push_back(&x);
+    }
+
+    template <typename U>
+    void push_back(const U& x)
```

Comment:
Is it wise to allow pushing anything (castable) into an array? I'm not convinced...